src/nbl/asset/ICPUImage.cpp

```diff
@@ -0,0 +1,319 @@
+			are filled with 0 value, we use flatten with tight buffer for the
+			texel copy & prefill if we don't have fully flatten input
+		*/
+
+		if (!isFullyFlatten) // TODO: we may think of even better optimization if we have mixed regions for mips (eg. a few flattened, some others not and left empty)
+		{
+			// create own regions & hook tight buffer with no gaps from scratch
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(parameters.mipLevels);
+			{
+				size_t bufferSize = 0ull;
+
+				for (auto rit = regions->begin(); rit != regions->end(); rit++)
+				{
+					auto miplevel = static_cast<uint32_t>(std::distance(regions->begin(), rit));
+					auto localExtent = state->inImage->getMipSize(miplevel);
+					rit->bufferOffset = bufferSize;
+					rit->bufferRowLength = localExtent.x; // could round up to multiple of 8 bytes in the future
+					rit->bufferImageHeight = localExtent.y;
+					rit->imageSubresource.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT; // otherwise won't pass validaiton 
+					rit->imageSubresource.mipLevel = miplevel;
+					rit->imageSubresource.baseArrayLayer = 0u;
+					rit->imageSubresource.layerCount = parameters.arrayLayers;
+					rit->imageOffset = { 0u,0u,0u };
+					rit->imageExtent = { localExtent.x,localExtent.y,localExtent.z };
+					auto levelSize = info.roundToBlockSize(localExtent);
+					auto memsize = levelSize[0] * levelSize[1] * levelSize[2] * parameters.arrayLayers * bytesPerPixel;
+
+					assert(memsize.getNumerator() % memsize.getDenominator() == 0u);
+					bufferSize += memsize.getIntegerApprox();
+				}
+			}
+
+			auto out = ICPUImage::create(IImage::SCreationParams(parameters));
+			out->setBufferAndRegions(core::smart_refctd_ptr(getScratchAsBuffer(scratch.flatten.size, scratch.flatten.offset)), std::move(regions));
+
+			auto& flatten = proxy.flatten;
+			flatten.inImage = state->inImage;
+			flatten.outImage = out;
+			flatten.preFill = true;
+			memset(flatten.fillValue.pointer, 0, sizeof(flatten.fillValue.pointer));
+
+			assert(CFlattenRegionsImageFilter::execute(policy, &proxy.flatten)); // this should never fail, at this point we are already validated
+		}
+
+		/*
+			now when the output is prepared we go with single
+			region covering all texels for a given mip level & layers,
+			note if we have optimized (fully flattened) input we skip
+			flattening by hand to avoid extra allocations & copies
+		*/
+
+		const auto* const image = isFullyFlatten ? state->inImage : proxy.flatten.outImage.get();
+		const auto texelOrBlockByteSize = asset::getTexelOrBlockBytesize(parameters.format);
+		const uint8_t* inData = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+		std::vector<uint32_t> layers(parameters.arrayLayers);
+		std::iota(layers.begin(), layers.end(), 0);
+
+		std::vector<uint32_t> levels(parameters.mipLevels);
+		std::iota(levels.begin(), levels.end(), 0);
```

Comment:
make your own custom iterators over a range of integers (someone probably has done it already), instead of using `iota`, the number of arrayLayers can be stupidly high